### PR TITLE
fix: Text updateTransform cannot be triggered before Render

### PR DESCRIPTION
**Pull request by JetLua**
fix: #6744 

fixed: https://jsfiddle.net/JetLu/w501jpxk/
```js
label.text += '100'
app.render()
```

Only in step `2` will `Text.updateText` be triggered, but it's too late.

`Text.updateText` -> `Text.updateTexture` -> `Sprite._onTextureUpdate` -> `Transform.onChange`

`Text` update `transform` property lags behind `render`.

**This PR may be only a temporary solution.**

image

### CODE REVIEW

**Comment by bigtimebuddy** [comment]
Could you check out the failing test?
https://github.com/pixijs/pixi.js/pull/6751/checks?check_run_id=847661289#step:7:1333

File: packages/text/src/Text.ts [hunk 1]
```diff
@@ -165,6 +165,11 @@ export class Text extends Sprite
     {
         const style = this._style;
 
+        if (!style)
```
Comment (bigtimebuddy) [inline packages/text/src/Text.ts:168]: I get the `updateText` in the `set text` method, but why the check for no style here?

Comment (JetLua) [reply packages/text/src/Text.ts:168]: image
`this.text` triggers the `set` before `this.style`

Comment (JetLua) [reply packages/text/src/Text.ts:168]: Maybe I can switch their places

**Comment by JetLua** [comment]
> Could you check out the failing test?
> https://github.com/pixijs/pixi.js/pull/6751/checks?check_run_id=847661289#step:7:1333

I changed the content of the `unit-test` because the `width` is `1` when `text` = `\n`

https://jsfiddle.net/JetLu/xnhftc31/

**Comment by bigtimebuddy** [comment]
Would it be possible to add a new unit test for this case? If we do find a better solution, would be good to prevent a regression.

**Comment by JetLua** [comment]
> Would it be possible to add a new unit test for this case? If we do find a better solution, would be good to prevent a regression.

hmm, there's no need to add new unit tests. I'm just correcting that value.

```js
const label = new PIXI.Text("\n");

label.updateText() // Without this line，output: 3

console.log(label.canvas.width) // output: 1
```

**Comment by bigtimebuddy** [comment]
Oh I get it! Awesome.

File: packages/text/src/Text.ts [hunk 2]
```diff
@@ -713,6 +713,7 @@ export class Text extends Sprite
         }
         this._text = text;
         this.dirty = true;
+        this.updateText(true);
```
Comment (themoonrat) [inline packages/text/src/Text.ts:716]: I am a bit wary of this change, as I think it'll bring performance issues for fast changing text.

Text in PixiJS has always been 'lazy'. So if you updated text a million times within 1 frame..... it doesn't matter, the texture for text only got updated when it comes through to render.
Now we're saying, every time you change the text, it's going to update that texture immediately.

That's a different performance profile which may negatively effect content out there - I'm thinking of high scores or numbers counting up super fast.

Comment (JetLua) [reply packages/text/src/Text.ts:716]: It is worth considering. Another option is to call `updateTransform` manually in `updateTexture`.

**Comment by JetLua** [comment]
🆕 : https://jsfiddle.net/JetLu/cpmsntq2/

**Comment by ivanpopelyshev** [comment]
need _recursivePostUpdateTransform or something like that

**Comment by JetLua** [comment]
> need _recursivePostUpdateTransform or something like that

3Q, updated

File: packages/text/src/Text.ts [hunk 3]
```diff
@@ -394,6 +394,9 @@ export class Text extends Sprite
 
         baseTexture.setRealSize(canvas.width, canvas.height, this._resolution);
 
+        // Recursively updates transform of all objects from the root to this one
+        (this as any)._recursivePostUpdateTransform();
```
Comment (bigtimebuddy) [inline packages/text/src/Text.ts:398]: Much better fix, however, can you please remove `as any`. Keeping it like this masks these API when we have to refactor things later.

The approach to deal with private APIs that are used internally, is to remove the access modifier (in this case `private`) from the original method. We can use documentation flagging to make sure it's not public. 

https://github.com/pixijs/pixi.js/blob/3f90364d02b11b4fef1645df2bedac3e9b2b121e/packages/display/src/DisplayObject.ts#L283

Comment (ivanpopelyshev) [reply packages/text/src/Text.ts:398]: yes,. this really should not be private

**Pull request merged**